src/extensions/crypt/mod-crypt.c

```diff
@@ -190,158 +189,158 @@ static REBNATIVE(rsa)
 {
     CRYPT_INCLUDE_PARAMS_OF_RSA;
 
-    REBYTE *n = NULL;
-    REBYTE *e = NULL;
-    REBYTE *d = NULL;
-    REBYTE *p = NULL;
-    REBYTE *q = NULL;
-    REBYTE *dp = NULL;
-    REBYTE *dq = NULL;
-    REBYTE *qinv = NULL;
-
-    REBINT n_len = 0;
-    REBINT e_len = 0;
-    REBINT d_len = 0;
-    REBINT p_len = 0;
-    REBINT q_len = 0;
-    REBINT dp_len = 0;
-    REBINT dq_len = 0;
-    REBINT qinv_len = 0;
-
-    REBCTX *obj = VAL_CONTEXT(ARG(key_object));
-
-    REBVAL *key = CTX_KEYS_HEAD(obj);
-    REBVAL *var = CTX_VARS_HEAD(obj);
-
-    for (; NOT_END(key); ++key, ++var) {
-        if (VAL_KEY_SYM(key) == SYM_SELF or IS_BLANK(var))
-            continue;
-
-        if (not IS_BINARY(var))
-            fail (Error(RE_EXT_CRYPT_INVALID_KEY_DATA, var, key, END));
-
-        REBSTR* word = VAL_KEY_CANON(key);
-        if (word == CRYPT_WORD_N) {
-            n = VAL_BIN_AT(var);
-            n_len = VAL_LEN_AT(var);
-        }
-        else if (word == CRYPT_WORD_E) {
-            e = VAL_BIN_AT(var);
-            e_len = VAL_LEN_AT(var);
-        }
-        else if (word == CRYPT_WORD_D) {
-            d = VAL_BIN_AT(var);
-            d_len = VAL_LEN_AT(var);
-        }
-        else if (word == CRYPT_WORD_P) {
-            p = VAL_BIN_AT(var);
-            p_len = VAL_LEN_AT(var);
-        }
-        else if (word == CRYPT_WORD_Q) {
-            q = VAL_BIN_AT(var);
-            q_len = VAL_LEN_AT(var);
-           break;
-        }
-        else if (word == CRYPT_WORD_DP) {
-            dp = VAL_BIN_AT(var);
-            dp_len = VAL_LEN_AT(var);
-        }
-        else if (word == CRYPT_WORD_DQ) {
-            dq = VAL_BIN_AT(var);
-            dq_len = VAL_LEN_AT(var);
-        }
-        else if (word == CRYPT_WORD_QINV) {
-            qinv = VAL_BIN_AT(var);
-            qinv_len = VAL_LEN_AT(var);
-        }
-        else
-            fail (Error(RE_EXT_CRYPT_INVALID_KEY_FIELD, key, END));
-    }
+    REBVAL *obj = ARG(key_object);
 
-    if (!n || !e)
-        fail (Error(RE_EXT_CRYPT_INVALID_KEY, ARG(key_object), END));
+    // N and E are required
+    //
+    REBVAL *n = rebRun("ensure binary! pick", obj, "'n", END);
+    REBVAL *e = rebRun("ensure binary! pick", obj, "'e", END);
 
```

Comment:
Wouldn't this cause quite some overhead? It needs to scan the string, and evaluate the block. Seems too much work for something that can be done by a pointer comparison in current implementation.

Reply:
I mentioned that such code segments could be factored into functions and then you use the function.  Which would take care of the scanning and binding, anyway.  That could be a first avenue of performance if this turned out to be slow.

When you consider especially the need to hit a network connection to grab data, it should (hopefully) be negligible by comparison.  Remember:

> 90% of a program's execution time is spent in only 10% of its code. The standard inference from this rule is that programmers should find that 10% of the code and optimize it ... But a second inference is just as important: programmers can deoptimize the other 90% of the code (in order to make it easier to use, maintain, etc.)

The previous code didn't check that it was a BINARY! even, before doing VAL_BIN_AT().  If somehow the object had bad data poked into it, it would crash.  Being able to express that as `ensure binary!` or `opt match binary!` (if you'd like to get null back rather than raise an error) has far greater leverage.

The Rebol codebase has simply too much code to reasonably tune all of at the native level.  While it's good to have the option, I see the choice of picking libRebol vs. managing series directly as a pragmatic one.  Attack the bottlenecks and make everything else robust.

Reply:
I would like to upvote your decision on this one HF.

Reply:
Certainly we can still have code that uses the internal API...when it's appropriate.  But all the extensions that were using it were ones that I think are better off being migrated off the internal API *completely*.  If we need them to be faster, I'd like to do that by finding subpatterns in what they do and optimizing *those* as natives.  So none of those are instances that would motivate me to keep it.

@zsx If you want the NEW-WORDS feature, it can be kept if it's rewritten to use API REBVALs.  Those REBVALs would need to get the same lifetime as the extension (one way or another, which today would mean unmanaging them and then releasing them in the extension's shutdown).  Then anyone with internal API can use VAL_WORD_CANON on them just like any other REBVAL.

If it's a priority for you for reasons beyond these three usages, then that is fine.  The issue just being that we have a *lot* of auto-generating code.  So if we take the total amount of code we have to deal with--factor in its maintenance and understanding--then compare to asking people to write:

      REBVAL x_word;

      ...Init_Extension... {
          x_word = rebUnmanage(rebRun("quote x", END));
      }

      ...Shutdown_Extension... {
          rebRelease(x_word);
      }

That might be a lot less "opaque" than trying to automate it.  Because there are many such optimizations one might think of, and maintaining invisible automation code has become a real headache.

Plus you never really know.  They may be generating these words, but every time they use them they wind up putting them in arrays.  Is the real feature not "new words" but "boilerplate"...like fragments of code that could be a word, or could be a BLOCK!, etc?

That's the kind of challenge I want us to be thinking of when we're looking at these big C macro generators.  To be careful when we make them, and be sure that when we do they're really buying us something over writing the code by hand.